File: SlidingWindow/LongestSubstringwithAtMostTwoDistinctCharacters.java
import java.util.HashMap;

public class LongestSubstringwithAtMostTwoDistinctCharacters {
    public static void main(String[] args) {
        LongestSubstringwithAtMostTwoDistinctCharacters solution = new LongestSubstringwithAtMostTwoDistinctCharacters();
        String s = "ccaabbb";
        int ans = solution.input(s);
        System.out.println(ans);
    }

    public int input(String s) {
        int left = 0;
        int maxLength = 0;
        HashMap<Character, Integer> charmap = new HashMap<>();
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            charmap.put(c, charmap.getOrDefault(c, 0) + 1);
            while (charmap.size() > 2) {
                char leftChar = s.charAt(left);
                charmap.put(leftChar, charmap.get(leftChar) - 1);
                if (charmap.get(leftChar) == 0) {
                    charmap.remove(leftChar);
                }
                left++;
            }
            maxLength = Math.max(maxLength, right - left + 1);
        }
        return maxLength;
    }
}

File: SlidingWindow/LongestSubstringwithAtMostTwoDistinctCharacters.js
var LongestSubstringwithAtMostTwoDistinctCharacters=function(s){
    let map=new Map();
    let start=0;
    let max=0;
    let i=0;
    for(i=0;i<s.length;i++){
      let c=s.charAt(i);
      map.set(c,(map.get(c)||0)+1);
      while(map.size>2){
        let startChar=s.charAt(start);
        map.set(startChar,map.get(startChar)-1);
        max=Math.max(max,i-start);
        if(map.get(startChar)===0){
            map.delete(startChar);
        }
        start++;
      }
      
    }
    max=Math.max(max,i-start);
    return max;
};
const s="ccaabbb";
const result=LongestSubstringwithAtMostTwoDistinctCharacters(s);
console.log(result);